### PR TITLE
Move wallet callbacks into cs_main (this effectively reverts #7946)

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This is an alternative to #9570 as well as at least two other fixes which would be required to wallet consistency issues generated by #7946, none of which I believe are reasonable to make 0.14.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2473,20 +2473,20 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
             pindexNewTip = chainActive.Tip();
             pindexFork = chainActive.FindFork(pindexOldTip);
             fInitialDownload = IsInitialBlockDownload();
+
+            // throw all transactions though the signal-interface
+            // while _not_ holding the cs_main lock
```
Comment (sipa) [inline src/validation.cpp]: This comments seems outdated now.

File: src/validation.cpp [hunk 2]
```diff
@@ -2473,20 +2473,20 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
...
+            for (const auto& pair : connectTrace.blocksConnected) {
+                assert(pair.second);
+                const CBlock& block = *(pair.second);
+                for (unsigned int i = 0; i < block.vtx.size(); i++)
+                    GetMainSignals().SyncTransaction(*block.vtx[i], pair.first, i);
+            }
         }
```
Comment (sipa) [inline src/validation.cpp]: I think you need to clear connectTrace.blocksConnected here.

Comment (TheBlueMatt) [inline src/validation.cpp]: It will be automatically when it continues the loop.

File: src/validation.cpp (hunk 1 above)
Comment (TheBlueMatt) [inline src/validation.cpp]: Fixed

**Comment by morcos** [comment]
utACK

I know we think this is on the top of our minds for further improvement, but I think it would make sense to document at least in the comments and maybe also the commit message why this matter.

**Comment by TheBlueMatt** [comment]
Fixes #9584 and #9148 temporarily, with a hope for reversion in 0.15.

**Comment by jonasschnelli** [comment]
utACK 989989354b68d3954fe2742b96c53eeb2e8a7670. I think it make sense to apply this revert and go for a better solution in 0.15.

**Comment by laanwj** [comment]
Concept ACK

**Comment by jtimon** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
Concept ACK

**Comment by gmaxwell** [comment]
Concept ACK.

File: src/validation.cpp (hunk 2 above)
Comment (sipa) [inline src/validation.cpp]: Oh, loop-scoped variable, never mind.

**Review by sipa — COMMENTED** [review]
utACK

**Comment by laanwj** [comment]
utACK 9899893

**Pull request merged**